lib/pacemaker/pcmk_verify.c

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright 2023 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+#include <crm/cib/internal.h>
+#include <crm/common/output.h>
+#include <crm/common/results.h>
+#include <crm/common/scheduler.h>
+#include <pacemaker-internal.h>
+#include <pacemaker.h>
+
+#include <stdint.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "libpacemaker_private.h"
+
+int
+parse_cib_file(pcmk__output_t *out, bool use_live_cib, const char *xml_file,
+             const char *xml_string, bool xml_stdin, xmlNodePtr *cib_object)
+{
+    int rc = pcmk_rc_ok;
+
+    crm_info("=#=#=#=#= Getting XML =#=#=#=#=");
+
+    if (use_live_cib) {
+        crm_info("Reading XML from: live cluster");
+        rc = cib__signon_query(out, NULL, cib_object);
+
+    } else if (xml_file != NULL) {
+        *cib_object = filename2xml(xml_file);
+        if (*cib_object == NULL) {
+            rc = ENODATA;
+        }
+
+    } else if (xml_string != NULL) {
+        *cib_object = string2xml(xml_string);
+        if (*cib_object == NULL) {
+            rc = ENODATA;
+        }
+
+    } else if (xml_stdin) {
+        *cib_object = stdin2xml();
+        if (*cib_object == NULL) {
+            rc = ENODATA;
+        }
+
+    } else {
+        rc = ENODATA;
+    }
+
+    if (rc == ENODATA) {
+        out->err(out, "Couldn't parse input file");
+    }
+
+    return rc;
+
+}
+
+int
+pcmk__verify(pcmk_scheduler_t *scheduler, pcmk__output_t *out,
+             bool use_live_cib, const char *cib_save, int verbosity, xmlNode *cib_object)
+{
+    int rc = pcmk_rc_ok;
+    xmlNode *status = NULL;
+
+    if (!pcmk__xe_is(cib_object, XML_TAG_CIB)) {
+        rc = EBADMSG;
+        out->err(out, "This tool can only check complete configurations (i.e. those starting with <cib>).");
+        goto verify_done;
+    }
+
+    if (cib_save != NULL) {
+        write_xml_file(cib_object, cib_save, FALSE);
+    }
+
+    status = pcmk_find_cib_element(cib_object, XML_CIB_TAG_STATUS);
+    if (status == NULL) {
+        create_xml_node(cib_object, XML_CIB_TAG_STATUS);
+    }
+
+    if (pcmk__validate_xml(cib_object, NULL, (xmlRelaxNGValidityErrorFunc) out->err, out) == FALSE) {
+        pcmk__config_err("CIB did not pass schema validation");
+        rc = pcmk_rc_schema_validation;
+        goto verify_done;
+
+    } else if (cli_config_update(&cib_object, NULL, FALSE) == FALSE) {
```

Comment:
I just noticed cli_config_update() takes a `to_logs` argument and prints to stderr if false. That means those errors won't be added to XML output. I'll create a separate project to handle output objects there.